### ISSUE TITLE
Stack overflow in debug mode in Asteroids::Asteroids

### ISSUE THREAD

**Issue by nlguillemot** [GameTechDev/asteroids_d3d12#3]
When I build and run the program in debug mode, it crashes at the start of Asteroids::Asteroids() due to a stack overflow.

I looked around and couldn't find the culprit (this usually happens due to large stack allocated arrays). I did notice that commenting out the `for (int j = 0; j < NUM_ASTEROIDS; ++j)` loop in the constructor makes the stack overflow go away, but that code doesn't contain any large stack allocations. Compiler bug?

Works fine in release mode.

Using VS 2017 community edition, version 15.3.1. Windows Build 15063.
(Update: Still happens in version 15.3.2)

**Comment by jdm3** [comment]
I could not reproduce this on my current setup: Win10 15063, VS pro 2017 15.1 26403.7

**Comment by jdm3** [comment]
Just grabbed vs community 15.3.4 and worked there too.

Was there anything else you changed?  Is there any more debugging you can do on your side?

**Comment by nlguillemot** [comment]
Still happening on VS Community 2017 15.3.4, also on Windows 10 15063. No local modifications, with a fresh pull from GitHub.

---

I did some debugging, and found something kind of interesting. Here's what I found:

1. Increasing the reserved stack size to 100MB (an arbitrary large amount) makes the crash go away.
2. Decreasing the number of asteroids to roughly less than 100 also makes the crash go away.

Based on these observations, I thought the following struct was kind of suspicious, since it could potentially be used in a big stack allocation somewhere:

```cpp
struct DynamicUploadHeap {
    DrawConstantBuffer mDrawConstantBuffers[NUM_ASTEROIDS];
    SkyboxConstantBuffer mSkyboxConstants;
    ExecuteIndirectArgs mIndirectArgs[NUM_ASTEROIDS];
    SpriteVertex mSpriteVertices[MAX_SPRITE_VERTICES_PER_FRAME];
};
```

I found that it wasn't used in any stack allocation, and instead it's used as an `UploadHeapT<DynamicUploadHeap>`, so that wasn't quite the problem.

Still, I thought there might be something weird about big structs, so I tried to replace that struct with manual memory allocations:

```cpp
// linearly allocate the dynamic upload heap data
UINT64 dynamicUploadHeapSize = 0;

// allocate draw constants
dynamicUploadHeapSize = (dynamicUploadHeapSize + alignof(DrawConstantBuffer)-1) & ~(alignof(DrawConstantBuffer)-1);
UINT64 drawConstantsDynUploadOffset = dynamicUploadHeapSize;
dynamicUploadHeapSize += sizeof(DrawConstantBuffer) * NUM_ASTEROIDS;

// allocate skybox constants
dynamicUploadHeapSize = (dynamicUploadHeapSize + alignof(SkyboxConstantBuffer)-1) & ~(alignof(SkyboxConstantBuffer)-1);
UINT64 skyboxConstantsDynUploadOffset = dynamicUploadHeapSize;
dynamicUploadHeapSize += sizeof(SkyboxConstantBuffer);

// allocate indirect args
dynamicUploadHeapSize = (dynamicUploadHeapSize + alignof(ExecuteIndirectArgs)-1) & ~(alignof(ExecuteIndirectArgs)-1);
UINT64 indirectArgsDynUploadOffset = dynamicUploadHeapSize;
dynamicUploadHeapSize += sizeof(ExecuteIndirectArgs) * NUM_ASTEROIDS;

// allocate sprite vertices
dynamicUploadHeapSize = (dynamicUploadHeapSize + alignof(SpriteVertex)-1) & ~(alignof(SpriteVertex)-1);
UINT64 spriteVerticesDynUploadOffset = dynamicUploadHeapSize;
dynamicUploadHeapSize += sizeof(SpriteVertex) * MAX_SPRITE_VERTICES_PER_FRAME;
```

I also demoted the `UploadHeapT<DynamicUploadHeap>` to a plain `UploadHeap`, and replaced all its memory accesses with manually offsetted pointers based on the offsets computed above.

To my surprise, the program was still crashing with a stack overflow. From there, I found the real culprit: `offsetof`. For some reason, using `offsetof` in `Asteroids::Asteroids` somehow causes the function to ask for a huge stack frame. Just replacing those uses with manually computed offsets solves the crash. It's not a problem to use `UploadHeapT` itself.

Further investigation might need insight into the compiler. I guess the compiler devs just don't expect people to declare structs with big arrays, even if the struct is heap allocated. It's strange that it doesn't happen on your computer though.

# tl;dr

Calling `offsetof` on large structs may cause a stack overflow, potential compiler bug.

Workaround: Replace uses of `offsetof` with the manually computed offsets as shown in the above snippet.

**Comment by jdm3** [comment]
Is "manually computed offsets" the code you show above?

Have you tried using the non-c++ or builtin version of offsetof()?

```cpp
#if defined(_MSC_VER) && !defined(_CRT_USE_BUILTIN_OFFSETOF)
    #ifdef __cplusplus
        #define offsetof(s,m) ((size_t)&reinterpret_cast<char const volatile&>((((s*)0)->m)))
    #else
        #define offsetof(s,m) ((size_t)&(((s*)0)->m))
    #endif
#else
    #define offsetof(s,m) __builtin_offsetof(s,m)
#endif
```

**Comment by nlguillemot** [comment]
>Is "manually computed offsets" the code you show above?

Yup.

>Have you tried using the non-c++ or builtin version of offsetof()?

I tested out those 3 different implementations of `offsetof`. The first two work correctly, and the third one (`__builtin_offsetof(s,m)`) crashes.

**Comment by jdm3** [comment]
OK, that makes more sense -- I was really confused how the ((s*)0)->m versions would cause a stack overflow!

On my build it directs to those because _MSC_VER is defined and _CRT_USE_BUILTIN_OFFSETOF isn't -- though, I'm not sure why that isn't the case for you.

**Comment by nlguillemot** [comment]
If I paste that code snippet as-is, it picks the first version, which doesn't crash. If I manually force it to call `__builtin_offsetof`, then it crashes. If I don't include that code snippet at all, it also crashes, since I guess Visual C++ uses something equivalent to `__builtin_offsetof` internally by default.

Another detail, to be precise. It's not the actual call to `__builtin_offsetof` that causes a stack overflow. The stack overflow happens before the body of `Asteroids::Asteroids` even begins, when its stack frame is allocated. I would guess that `__builtin_offsetof` somehow leads the compiler to allocate a huge stack frame.

**Comment by jdm3** [comment]
That code snipit comes directly from "c:\Program Files (x86)\Windows Kits\10\Include\10.0.15063.0\ucrt\stddef.h" on my machine, which is where my compiler is finding the offsetof definition.

Do you know where your compile is finding offsetof() from?

**Comment by nlguillemot** [comment]
Ok weird, I see that same definition of `offsetof` in `stddef.h` if I click on "Go To Definition".

I was wrong, the syntax highlighter tells me that the first version is getting chosen, but it actually picks the third one.

According to the test below, `_CRT_USE_BUILTIN_OFFSETOF` is defined in my build, which explains why it picks the third option.

```cpp
#ifdef _CRT_USE_BUILTIN_OFFSETOF
static_assert(false, "_CRT_USE_BUILTIN_OFFSETOF");
#endif
```

I don't know why that would be the case though. I didn't change any settings, and that macro isn't set in the project settings.

Just found out it looks like a known recent issue: https://developercommunity.visualstudio.com/content/problem/95908/offsetof-causing-stack-overflow-in-debug-builds.html

**Comment by jdm3** [comment]
I'd guess it's due to the different ways we each ended up at 15.3.4, but I'm not sure.  It doesn't seem to be defined for me.  If you can figure out the right place to force undefine it I'd be happy to pull that change back in.

**Comment by nlguillemot** [comment]
I think that the addition of `_CRT_USE_BUILTIN_OFFSETOF` by the VS devs was deliberate, since apparently reinterpret_cast is illegal in constexpr functions, but people want to use offsetof in constexpr functions. I think we can just wait for them to fix the compiler bugs, and use this issue as a reference workaround if it's a problem in the meantime.

**Comment by DaveBookout-Intel** [comment]
this was fixed in October 2017 release
